app/code/Magento/PageCache/Model/HTTP/PhpEnvironment/ResponsePlugin.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace Magento\PageCache\Model\HTTP\PhpEnvironment;
+
+use Magento\Framework\HTTP\PhpEnvironment\Response as Subject;
+use Magento\PageCache\Model\Config;
+use Zend\Http\HeaderLoader;
+use Magento\PageCache\Model\HTTP\Header\XMagentoTags;
+
+class ResponsePlugin
+{
+    /**
+     * Approximately 8kb in length
+     *
+     * @var int
+     */
+    private $requestSize = 8000;
+
+    /**
+     * PageCache configuration
+     *
+     * @var Config
+     */
+    private $config;
+
+    /**
+     * ResponsePlugin constructor.
+     *
+     * @param Config $config
+     */
+    public function __construct(
+        Config $config
+    ) {
+        $this->config = $config;
+    }
+
+    /**
+     * Special case for handling X-Magento-Tags header
+     * splits very long header into multiple headers
+     *
+     * @param Subject  $subject
+     * @param \Closure $proceed
+     * @param string   $name
+     * @param string   $value
+     * @param bool     $replace
+     *
+     * @return Subject|mixed
+     */
+    public function aroundSetHeader(Subject $subject, \Closure $proceed, $name, $value, $replace = false)
```

Comment:
It's not recommended to add new around plugins, as it has a greater performance penalty, specifically, in such a critical part of the system.